### ISSUE TITLE
New GoCD agents - "A token has already been issued for this agent"

### ISSUE THREAD

**Issue by Diogomrol** [gocd/gocd#5170]
##### Issue Type
- Bug Report

##### Summary

Having a GoCD Server with version 18.9.0 and adding a set if GoCD agents with version 18.5.0, after allocating resources to the agents and they start running a job it then goes 'Lost Contact' on some of them. Observing the logs we see a lot of messages about "A token has already been issued for this agent". This doesn't happen always, it seems to only happen sometimes.

This behavior also happened before when we had GoCD Server with 18.5.0 and an agent with 17.10.0.

This is bad for us because we have machine templates that come with the version 18.5.0 of GoCD agent and we were relying on GoCD automatic upgrade to put the agent in 18.9.0. This way we aren't being able to leverage this capability.

Agent log: [go-agent.log](https://github.com/gocd/gocd/files/2383390/go-agent.log)

###### Basic environment details

* Go Version: 18.9.0 (agent with 18.5.0)
* JAVA Version: 8
* OS: Windows

##### Steps to Reproduce

1. Have a GoCD server in version 18.9.0
2. Enable a set of new agents that have GoCD agent with version 18.5.0
3. Add resources / environments to the agents
4. Have a job start running on them
5. Sometimes some of them go to state LostContact with token errors

Thank you.

Regards,
Diogo

**Comment by Diogomrol** [comment]
@ketan, to give you more context on this:

We are creating our templates using infrastructure as code, and the last step of the template generation is installing the GoCD agent. 

We are installing GoCD using the command line silent installation with the following parameters:

`/S /START_AGENT=NO /SERVERURL=`"$GoServerURL`" /D=C:\Go\GoAgent`

So, this means that the agent shouldn't start after installation (which should not create the guid.txt file and so on, correct?).

After that, we then instantiate this template for multiple machines, so all of them should come without any file generated in the config folder.

Thank you :)

**Comment by Diogomrol** [comment]
Hi people,

We have some news on this: it looks like this doesn't have anything to do with the upgrade. Even if the agents are on the same version this is happening sometimes.

**Comment by ankitsri11** [comment]
Your understanding is correct, using command line silent installation, it shouldn't create guid.txt, token and agent.jks files. You should only see wrapper-agent.conf file under config folder. 

Do you see any message after installing GoCD agent using the command line silent installation? How do you register these agents? manually or using auto-register property file? 

Can explain a bit more about this - "After that, we then instantiate this template for multiple machines, all of them should come without any file generated in the config folder."

**Comment by sverrewl** [comment]
I'm seeing the same issue. We have stateless agents, built from the ground up on each boot. Creates guid.txt based on 'hostname -s' before starting the agent. Still, the server is refusing the agent with the issue described. For more detailed logs ( which includes java stack traces ) - please feel free look at the related logs from server and agent shared here https://www.dropbox.com/s/xlvzrmos1nzw3tq/gocd_token_issues_5170.tar.gz?dl=0

**Comment by ketan** [comment]
We're trying to reproduce this at our end. Could you confirm if you're bringing up multiple agents at around the same time? If yes, then could you set the log level to `debug` for the package `com.thoughtworks.go.agent.service` before bringing up the agent?

You can find out more about enabling and configuring enabling logging here https://docs.gocd.org/current/advanced_usage/logging.html#gocd-agent

**Comment by Diogomrol** [comment]
On our side the problem seems to happen when bringing up multiple agents at the same time, and on these situations this happens in some of them (sometimes in 1 or 2, other times in almost all of them).

**Comment by ketan** [comment]
In either case — would you mind turning on the log level and capturing the logs, both on the server and agent. We'd like to get to the root of this, and the logs would help immensely.

Thanks.

**Comment by sverrewl** [comment]
Seems to be several options to the log. Why don't you just provide me with the xml files so it logs what you need.

**Comment by ketan** [comment]
On your agent, put this in your `config/agent-logback-include.xml`:

```xml
<included>
  <logger name="com.thoughtworks.go.agent.service" level="DEBUG"/>
</included>
```

On your server, put this in your `logback-include.xml` (either in `/etc/go`, or `c:\program files\go server\config\` as may be the case):

```xml
<included>
  <logger name="com.thoughtworks.go.server.controller" level="DEBUG"/>
</included>
```

**Comment by arvindsv** [comment]
> Creates guid.txt based on 'hostname -s' before starting the agent. Still, the server is refusing the agent with the issue described.

@ketan Won't this message show up in precisely this scenario? As far as I understand, the `token` file is a way for a GoCD server to connect an agent guid to a known agent (using the token it has provided and stored on the server side). This is to prevent a rogue agent from just knowing the (easily findable) guid and impersonating the agent. The token file is provided only once, during registration. So, rogue agents cannot use the same guid.

In this case, it looks like an agent is recreated and the guid is recreated (**I assume it's the same every time**) using the hostname. So, for all practical purposes, the server is rejecting an agent with the same guid, which does not have the token (since it is recreated). This seems like it's doing exactly what it is meant to.

The correct way to recreate agents is to use a different guid and use auto registration, or capture both the guid and the token before you recreate it and reuse _both_ those files.

**Comment by ketan** [comment]
> @ketan Won't this message show up in precisely this scenario?

I'm also looking at https://github.com/gocd/gocd/issues/5170#issuecomment-421360144 where it appears that the guid.txt is not present in the template.

**Comment by sverrewl** [comment]
@arvidsv : Yes, you have understood.

What can I say. It used to work without any problems before the upgrade. If I now need to keep the token then I will have to make changes. Please advice.

**Comment by arvindsv** [comment]
@sverrewl Yes. You will need to keep the token file around as well. Try it. Or, use auto registration.

**Comment by arvindsv** [comment]
@sverrewl This was changed last year, in November. https://www.gocd.org/releases/#17-12-0

**Comment by sverrewl** [comment]
Ah. That makes sense - last upgrade was Nov 21. Thank you !

**Comment by Matrikss** [comment]
Hi, I'm a colleague of @Diogomrol 
We didn't yet manage to activate the higher log level.

However, we have been trying to get more concrete information of the manifestation of this ongoing problem:
- Our agents are instantiated from a machine image that doesn't have neither the guid nor token files (the agent doesn't start at the time of the image creation).
- Our agents don't have auto registration.
- When the images are instantiated (usually multiple at the same time), the agent runs for the first time and connects to the server and generates the guid and token files.
- After we enable the agent and assign resources and environments **some agents lose the token file** and appear as Lost Contact.

Is the guid/token generation somehow connected/seeded to the machine hostname?
I ask this because when we provision our agents there is a time where multiple machines still have the hostname of the machine image before we change its hostname, so the first run of multiple agents is for the same hostname.

**Comment by wildthing81** [comment]
is there any resolution to the above problem. we are also facing similar issue with our GoCD installation (18.9.0)

**Comment by maheshp** [comment]
@Matrikss the guid generation is not based on the machine hostname. On agent startup in absence of guid.txt a random UUID is generated and stored in the guid.txt. 

From GoCD 17.12.0 agent requires a token from server to register with the server. Before making a request to registration an agent makes a call to the server to fetch a token by passing its guid, the token received is then stored in the token file and subsequently used for registration.

Ideally you should get the error `A token has already been issued for this agent` when you have multiple agents with the same guid. Or if the token file is deleted for a pending agent. 

Your server logs should have a message similar to `AgentRegistrationController:324 - Rejecting request for token. Error: HttpCode=[409] Message=[A token has already been issued for this agent.] Pending=[true/false] UUID=[some-uuid]` . Can you verify if there is an agent corresponding to the given UUID in the logs, you can verify that by visiting the page `http://<your-go-server-url>/go/agents/some-uuid`

**Comment by Tecla** [comment]
I have a similar case (I think), but in my case I have a Linux gocd server, and a windows laptop that is running an agent.  The windows agent appears to occasionally lose connection to the server, and when it reconnects it starts giving me the "A token has already been issued for this agent." message.

The difference?  I'm connecting to a VPN and disconnecting, and the server is on my local network (not on the VPN).  I have local routes set up so that the agent can still reach the server even on VPN.  I have another agent running in Linux (in a VM on the server) and it never has any trouble with the token.

Is it possible that the network config is used in the token generation, and can cause it to reject the agent due to network changes like that?

**Comment by ketan** [comment]
I've looked at the code (once again) and I'm unsure why this error exists, everything seems to point to a case where an agent uuid was possibly recycled (most likely unintentionally) and the UUID does not correspond to the token that is stored on the agent.

> Is it possible that the network config is used in the token generation, and can cause it to reject the agent due to network changes like that?

I don't believe so. Looking at the code, the HMAC is generated in the snippet below. As you can see the HMAC is initialized with the `tokenGenerationKey` attribute of the `<server/>` of your config xml: https://github.com/gocd/gocd/blob/880694aef2e4f17cb40d0bf490f650ab06f52fb0/server/src/main/java/com/thoughtworks/go/server/controller/AgentRegistrationController.java#L93-L104

The agent, before registration will ask for a token before registration. This token is generated using the HMAC (from above) and the uuid submitted by the agent: https://github.com/gocd/gocd/blob/880694aef2e4f17cb40d0bf490f650ab06f52fb0/server/src/main/java/com/thoughtworks/go/server/controller/AgentRegistrationController.java#L320-L335

The agent upon receiving the token, will submit it during registration. The token is then validated when an agent attempts to register: https://github.com/gocd/gocd/blob/880694aef2e4f17cb40d0bf490f650ab06f52fb0/server/src/main/java/com/thoughtworks/go/server/controller/AgentRegistrationController.java#L207-L232

**Comment by Tecla** [comment]
My case sounds different than this issue, then.  It's the same directory, same guid.txt, etc and when I try to restart the agent it just refuses to accept it.  I've been able to delete the config directory and restart the agent, and it will get a new UUID and then I can take it out of pending status on the server and it works fine.  Sometimes it works, sometimes not, and after restarting the agent (or the entire machine) it will allow the agent to register.  I'll try to get some debug logs and catch it in the act, and maybe make a new issue as it may be different than this one.

**Comment by ketan** [comment]
> I'll try to get some debug logs and catch it in the act.

Thank you. If you're able to reproduce this with certainty, we can definitely take a look at fixing this.

**Comment by fabriciocduarte** [comment]
Is there a solution or workaround for this issue? We have several agents that gets scrapped and re-provisioned fresh from time to time, during the provisioning we populate the guid.txt file with the agents guid from when it was first provisioned.

**Comment by arvindsv** [comment]
@fabriciocduarte The last sentence in [this comment](https://github.com/gocd/gocd/issues/5170#issuecomment-424671803) is what you need. It's not a workaround. It's just that the `token` file is probably deleted during reprovisioning, and it needs to be kept.

**Comment by arvindsv** [comment]
Closing as stale.
